File: src/send/recipients.ts
/**
 * Dev/test guard: only one recipient allowed unless ZMAIL_SEND_PRODUCTION is set.
 */

export const DEV_SEND_ALLOWLIST = "lewiscirne+zmail@gmail.com";

/** Extract bare email from "Name <addr@x>" or "addr@x". */
export function extractEmailAddress(raw: string): string {
  const trimmed = raw.trim();
  const angle = trimmed.match(/<([^>]+)>/);
  if (angle) return angle[1].trim();
  return trimmed;
}

function normalizeForCompare(addr: string): string {
  return extractEmailAddress(addr).toLowerCase();
}

/**
 * Throws if any recipient is not the dev allowlist and production mode is off.
 */
export function assertSendRecipientsAllowed(
  addresses: string[],
  env: NodeJS.ProcessEnv = process.env
): void {
  if (env.ZMAIL_SEND_PRODUCTION === "1" || env.ZMAIL_SEND_PRODUCTION === "true") {
    return;
  }
  const allowed = normalizeForCompare(DEV_SEND_ALLOWLIST);
  for (const addr of addresses) {
    if (!addr.trim()) continue;
    if (normalizeForCompare(addr) !== allowed) {
      throw new Error(
        `Send blocked: recipient "${addr}" is not allowed in dev/test. Only ${DEV_SEND_ALLOWLIST} is permitted, or set ZMAIL_SEND_PRODUCTION=1 to send to other addresses.`
      );
    }
  }
}

File: rust/src/send/recipients.rs
//! `ZMAIL_SEND_TEST` recipient allowlist.

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SendTestMode {
    Off,
    On,
}

pub fn parse_send_test_mode() -> SendTestMode {
    match std::env::var("ZMAIL_SEND_TEST") {
        Ok(v) if v == "1" || v.eq_ignore_ascii_case("true") => SendTestMode::On,
        _ => SendTestMode::Off,
    }
}

/// When send-test is on, only these addresses are allowed (mirrors TS allowlist shape).
pub fn filter_recipients_send_test(
    mode: SendTestMode,
    recipients: &[String],
    allowlist: &[String],
) -> Result<Vec<String>, String> {
    if mode == SendTestMode::Off {
        return Ok(recipients.to_vec());
    }
    let allow: std::collections::HashSet<String> =
        allowlist.iter().map(|s| s.to_lowercase()).collect();
    let mut out = Vec::new();
    for r in recipients {
        let l = r.to_lowercase();
        if allow.contains(&l) {
            out.push(r.clone());
        } else {
            return Err(format!(
                "ZMAIL_SEND_TEST=1: recipient {r} is not in the allowlist"
            ));
        }
    }
    Ok(out)
}
